src/simd/allocator.h

```diff
@@ -0,0 +1,165 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @file allocator.h
+ *
+ * Provide a custom memory allocator for SIMD.
+ *
+ * This allocator always returns memory that is suitably aligned to be loaded
+ * efficiently by a Register object.
+ */
+
+#ifndef __QUAD_SIMD_SIMD_ALLOCATOR_H__
+#define __QUAD_SIMD_SIMD_ALLOCATOR_H__
+
+#include <cassert>
+#include <cstdint>
+#include <limits>
+
+#include "simd/definitions.h"
+
+namespace quadiron {
+namespace simd {
+
+/// Check that the given address is properly aligned.
+template <typename T>
+inline bool addr_is_aligned(const T* addr)
+{
+    // Without SIMD, there is no specific alignment constraint.
+    if (INSTRUCTION_SET == InstructionSet::NONE) {
+        return true;
+    }
+    const std::uintptr_t address = reinterpret_cast<std::uintptr_t>(addr);
+    return (address & (ALIGNMENT - 1)) == 0;
+}
+
+/** Custom allocator to take advantage of SIMD processing.
+ *
+ * This allocator always return memory that is suitably aligned for the current
+ * SIMD instruction set. Thanks to this property, you can safely use the aligned
+ * load from the Register class in order to increase performance.
+ */
+template <typename T>
+class AlignedAllocator {
+  public:
+    using value_type = T;
+
+    AlignedAllocator() noexcept {}
+    // No state, => nothing to copy.
+    template <class U>
+    AlignedAllocator(AlignedAllocator<U> const& other) noexcept
+    {
+    }
+
+    value_type* allocate(std::size_t count)
+    {
+        // Guard against overflow!
+        if (count > max_size()) {
+            throw std::bad_alloc();
+        }
+
+        // No SIMD: default allocator is good enough!
+        if (INSTRUCTION_SET == InstructionSet::NONE) {
+            return static_cast<value_type*>(
+                ::operator new(count * sizeof(value_type)));
+        }
+
+        // Overallocate just enough to have room for alignment adjustment.
+        const std::size_t size = count * sizeof(value_type) + ALIGNMENT;
+        unsigned char* ptr = static_cast<unsigned char*>(::operator new(size));
+
+        // Align the allocated memory.
+        const std::uintptr_t address = reinterpret_cast<std::uintptr_t>(ptr);
+        const unsigned offset = ALIGNMENT - (address % ALIGNMENT);
+        assert(offset >= 1); // We need a byte to store the offset itself.
+        unsigned char* aligned_ptr = ptr + offset;
+
+        // Store the offset just before the aligned memory.
+        assert(offset <= std::numeric_limits<unsigned char>::max());
+        *(aligned_ptr - 1) = static_cast<unsigned char>(offset);
+
+        // Return the aligned pointer.
+        //
+        // Clang analyser think that we leak `ptr`, whereas we can re-compute
+        // it from `aligned_ptr` and free it in `deallocate`.
+        // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDeleteLeaks)
+        return reinterpret_cast<value_type*>(aligned_ptr);
+    }
+
+    void deallocate(value_type* ptr, std::size_t count) noexcept
```

Comment:
the 2nd argument is useless, isn't it?

Reply:
It's unused in my implementation, yes.
That being said it's specified like this (see [the specification of the method](https://en.cppreference.com/w/cpp/memory/allocator/deallocate)), and it can be used some allocator to improve the performance (example [here](https://github.com/jemalloc/jemalloc/issues/28)).

Reply:
so are we going to use this variable?

Reply:
No, we won't use it.
But it has to be here anyway, it's part of the allocator interface: if you don't have it then you're not implementing an allocator.

That being said, it's not an issue: the standard says "if you pass another value than the one used to allocate the behavior is undefined", well "undefined" can be anything including "it works" and since we are implementing it here, we get to decide how the code behave: and as you can see, you can give any value you want it will works (because we don't use it), nevertheless it has to be there to fullfill the requirement.

Reply:
Okay, so to benefit this advantage, should we use allocation size, e.g. `size = count * sizeof(value_type) + ALIGNMENT` for SIMD, in `::operator delete()` at line 120 & 131?

Reply:
Ha indeed, since C++14 the "standard" `delete` also accept a size.

I can make this change (but then, giving a random size to `deallocate` will really trigger an undefined behavior!).

Reply:
Hum, the `void operator delete  ( void* ptr, std::size_t sz );` seems not supported by Clang 6 (but works with GCC 8.2), so I think I'll revert it because we can't lost the ability to compile with Clang for an hypothetical performance improvement.

Reply:
Okay, so we don't do that. Anyway, we do no benefit advantage of sized deallocation, don't we?

Reply:
Anyway, I think the standard allocator from the libc/libc++ doesn't take advantage of it yet so yeah: no benefit for now I think (but we still have to keep the prototype).